src/Network/HTTP/Affjax.purs

```diff
@@ -117,6 +123,41 @@ delete u = affjax $ defaultRequest { method = DELETE, url = u }
 delete_ :: forall e. URL -> Affjax e Unit
 delete_ = delete
 
+-- | Retry a request with exponential backoff, timing out optionally after a specified number of milliseconds.
+retry :: forall e a b. (Requestable a) => Maybe Int -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b) -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b)
+retry milliseconds run req = do
+  failureVar <- makeVar
+  let loop = go failureVar
+  case milliseconds of
+    Nothing -> loop 1
+    Just milliseconds -> do
+      respVar <- makeVar
+      loopHandle <- forkAff $ loop 1 >>= putVar respVar <<< Just
+      timeoutHandle <-
+        forkAff <<< later' milliseconds $ do
+          putVar respVar Nothing
+          loopHandle `cancel` error "Cancel"
+      result <- takeVar respVar
+      case result of
+        Nothing ->
+          takeVar failureVar
+        Just resp -> pure resp
+  where
+    assert200 resp =
+      case resp.status of
+        StatusCode 200 -> Right resp
+        _ -> Left resp
+
+    go failureVar n = do
+      result <- run req
+      case assert200 result of
```

Comment:
Hmm, well, if an exception is thrown for a host-not-found type of error which can occur during transient connection difficulty, then we need to catch it. Otherwise, I probably would not catch it but make sure it ripples up. Seem right, @garyb?


-- 
John A. De Goes
Follow me @jdegoes

On August 3, 2015 at 5:13:35 PM, Gary Burgess (notifications@github.com) wrote:

In src/Network/HTTP/Affjax.purs:

> +          putVar respVar Nothing
> +          loopHandle `cancel` error "Cancel"
> +      result <- takeVar respVar
> +      case result of
> +        Nothing ->
> +          takeVar failureVar
> +        Just resp -> pure resp
> +  where
> +    assert200 resp =
> +      case resp.status of
> +        StatusCode 200 -> Right resp
> +        _ -> Left resp
> +
> +    go failureVar n = do
> +      result <- run req
> +      case assert200 result of
Exceptions are thrown, but unfortunately I've not been able to find an exact description of the conditions under which that happens (it's whenever there's an XHR onerror)... but generally it's when there's no way of getting a response at all (invalid URLs, etc). I'm not sure if that extends to requesting things from non-existant hosts however.

—
Reply to this email directly or view it on GitHub.